Emby.Server.Implementations/LiveTv/TunerHosts/HdHomerun/HdHomerunUdpStream.cs

```diff
@@ -131,33 +135,30 @@ public override async Task Open(CancellationToken openCancellationToken)
             await taskCompletionSource.Task.ConfigureAwait(false);
         }
 
-        private Task StartStreaming(UdpClient udpClient, HdHomerunManager hdHomerunManager, IPAddress remoteAddress, TaskCompletionSource<bool> openTaskCompletionSource, CancellationToken cancellationToken)
+        private async void StartStreaming(UdpClient udpClient, HdHomerunManager hdHomerunManager, IPAddress remoteAddress, TaskCompletionSource<bool> openTaskCompletionSource, CancellationToken cancellationToken)
```

Comment:
Never use async void if it can be avoided.

Reply:
That's true, however in this case it's needed as "StartStreaming" needs to run as a background process. It broke whenever not doing something with a Task was changed from a warning to an error and the person fixing it erroneously awaited on it.

Note that this pattern is used in numerous other places in the codebase as well for functions that should run as background tasks.

Reply:
FYI: async void should only be used with events.

It is used in some areas of the source incorrectly, and these are being slowly changed - as the code is updated.

Reply:
Had a look at the code. If i'm reading it correctly, StartStreaming is executed with the main thread waiting until the copy is complete. 

Couldn't the taskCompletionSource object be removed entirely, with the open function returning the async StartStreaming function? Something like...

```
        public override async Task Open(CancellationToken openCancellationToken)
        {
            ...
            catch (Exception ex)
            {
                using (udpClient)
                using (hdHomerunManager)
                {
                    if (!(ex is OperationCanceledException))
                    {
                        Logger.LogError(ex, "Error opening live stream:");
                    }

                    throw;
                }
            }

            MediaSource.Path = _appHost.GetLoopbackHttpApiUrl() + "/LiveTv/LiveStreamFiles/" + UniqueId + "/stream.ts";
            MediaSource.Protocol = MediaProtocol.Http;

            return awit StartStreaming(
                udpClient,
                hdHomerunManager,
                remoteAddress,
                taskCompletionSource,
                LiveStreamCancellationTokenSource.Token).ConfigureAwait(false);
        }
```

Reply:
@BaronGreenback StartStreaming is really a bad name for what this function actually does. It should really be called something like just "Stream" or "StreamThread" or "RunStreaming", etc., as it runs for as long as the stream runs; it takes any data sent to it from the HD Home Run and writes it to a file, presumably for ffmpeg to do whatever it needs to do for streaming. Returning the task from Open just pushes the problem one level higher (to LiveTvManager.GetChannelStream, where it's awaited on) when it should never be awaited on, as the API call that starts all of this needs to ultimately return for the stream to even start.

All I want here is for HD Home Run streaming to work... I don't want to make arguments about what a better way is to do something vs. another way, nor do I want to re-architect the application to work the "right" way; I just want this to work, and all I've done is made the minimal code changes necessary to restore functionality that was broken several years ago due to haphazard changes that were never tested.

If the change that I've made has broken something in some way, sure please let me know and I'll fix it, and if you want me to make some small specific code change to get this accepted and merged while also not breaking this again, sure, however this is getting to the point where it'll be easier for me to just maintain my own copy with this fix included and build the docker image myself, so I'm not willing to put much more effort into this.

Thank you.

Reply:
@ryanpetris we all want JF to work, that's why we're here, and thank for your contribution. 
Believe me, when I say that I know these reviews can be annoying and frustrating, but their purpose is to ensure that the next Dev that comes along and makes a change doesn't break anything, (like was done several years ago with his bit). The comments above aren't meant to be criticism, but are meant to aid in achieving this goal. And it's not just you that has to go through this process, we all do.

The good news is that you've fixed the problem, with the goal of keep the change to a minimum - something I'm yet to learn - and if you wish to maintain your own copy, then you are free to, as this is one of the possibilities with open source.

However, by going through this process, you will not only be helping with this change, but will be contributing to the project and aiding fellow HD users.

Please also be aware that this process isn't one that you go through alone - there are numerous developers who can provide help and offer assistance if requested.

Reply:
@BaronGreenback I understand your position, however please understand mine as well -- you and cvium have pointed out a perceived problem (one that I don't agree with) with a particular method signature (async void) even though said pattern is used in many many other places throughout the codebase, and neither has anyone provided what would be a satisfactory solution to the problem or pointed out where this has been resolved in other areas of the codebase.

As I have said before, this piece of code needs to run in the background, and needs to remain running for the duration of the livestream. We could start up a separate thread for it however given that it uses async/await within the function it's just going to end up back in the general task pool anyway, with now another separate thread running waiting for its completion. A similar problem would happen with a separate thread pool.

> ...but their purpose is to ensure that the next Dev that comes along and makes a change doesn't break anything, (like was done several years ago with his bit).

IMO if you're going to be refactoring code or doing things like fixing warnings, you really need to know what you're doing. The person that added the await in the first place obviously did not know what they were doing and merely followed the compilers suggestion of adding an await.

That said, given that the function signature is now "async void" rather than "async Task", it should be obvious to any future devs that this function is both meant to be run in the background and not awaited on.

So, as I've said, if you or someone would like to propose a solution that would remove the necessity of async void as well as _not_ require it to be awaited on (again, it needs to run in the background), then I'd be happy to update my code as well as test it to ensure it still works. However if all I'm going to get is conjecture over something that exists in many other places in the codebase without a proposal of even a somewhat viable solution, then I'll pass given that this is a voluntary contribution.

Reply:
@BaronGreenback and/or @cvium: @PrplHaz4 pointed out that in SharedHttpStream.cs, a similar operation is returning a Task but that task is just being discarded. Would this resolve your issues with async void?

https://github.com/jellyfin/jellyfin/pull/4182#issuecomment-712934715

Reply:
does for me

Reply:
@ryanpetris Task.Run is another can of worms, but it's fine in this case.

Reply:
This has been addressed.